py/dml/codegen.py

```diff
@@ -2034,74 +2055,75 @@ def stmt_for(stmt, location, scope):
     posts = codegen_statements(posts, location, scope)
-    loop_stack.append('c')
-    try:
+    with CLoopContext():
         res = mkFor(stmt.site, pres, cond, posts,
                     codegen_statement(statement, location, scope))
-    finally:
-        loop_stack.pop()
     return [res]
 
 @statement_dispatcher
 def stmt_dowhile(stmt, location, scope):
     [cond, statement] = stmt.args
     cond = as_bool(codegen_expression(cond, location, scope))
-    loop_stack.append('c')
-    try:
+    with CLoopContext():
         res = mkDoWhile(stmt.site, cond,
                         codegen_statement(statement, location, scope))
-    finally:
-        loop_stack.pop()
     return [res]
 
 @statement_dispatcher
 def stmt_switch(stmt, location, scope):
     [expr, body_ast] = stmt.args
     expr = codegen_expression(expr, location, scope)
-    loop_stack.append('c')
-    if stmt.site.dml_version() != (1, 2):
-        assert body_ast.kind == 'compound'
-        [stmt_asts] = body_ast.args
-        stmts = codegen_statements(stmt_asts, location, scope)
-        if (not stmts
-            or not isinstance(stmts[0], (ctree.Case, ctree.Default))):
-            raise ESWITCH(
-                body_ast.site, "statement before first case label")
-        defaults = [i for (i, sub) in enumerate(stmts)
-                    if isinstance(sub, ctree.Default)]
-        if len(defaults) > 1:
-            raise ESWITCH(stmts[defaults[1]].site, "duplicate default label")
-        if defaults:
-            for sub in stmts[defaults[0]:]:
-                if isinstance(sub, ctree.Case):
-                    raise ESWITCH(sub.site,
-                                  "case label after default label")
-        body = ctree.Compound(body_ast.site, stmts)
-    else:
-        body = codegen_statement(body_ast, location, scope)
-    try:
+    with CLoopContext():
+        if stmt.site.dml_version() != (1, 2):
+            assert body_ast.kind == 'compound'
+            [stmt_asts] = body_ast.args
+            stmts = codegen_statements(stmt_asts, location, scope)
+            if (not stmts
+                or not isinstance(stmts[0], (ctree.Case, ctree.Default))):
+                raise ESWITCH(
+                    body_ast.site, "statement before first case label")
+            defaults = [i for (i, sub) in enumerate(stmts)
+                        if isinstance(sub, ctree.Default)]
+            if len(defaults) > 1:
+                raise ESWITCH(stmts[defaults[1]].site,
+                              "duplicate default label")
+            if defaults:
+                for sub in stmts[defaults[0]:]:
+                    if isinstance(sub, ctree.Case):
+                        raise ESWITCH(sub.site,
+                                      "case label after default label")
+            body = ctree.Compound(body_ast.site, stmts)
+        else:
+            body = codegen_statement(body_ast, location, scope)
+
```

Comment:
Nice, this looks like a bugfix, ESWITCH could push the loop_stack out of sync.

Reply:
Yeah, I figured the old `try`-`finally` was weird.